libs/parallelism/execution/include/hpx/execution/algorithms/bulk.hpp

```diff
@@ -0,0 +1,256 @@
+                typename std::iterator_traits<shape_iterator>::value_type;
+
+            template <typename R_, typename Shape_, typename F_>
+            bulk_receiver(R_&& r, Shape_&& shape, F_&& f)
+              : r(std::forward<R_>(r))
+              , shape(std::forward<Shape_>(shape))
+              , f(std::forward<F_>(f))
+            {
+            }
+
+            template <typename E>
+            void set_error(E&& e) && noexcept
+            {
+                hpx::execution::experimental::set_error(
+                    std::move(r), std::forward<E>(e));
+            }
+
+            void set_done() && noexcept
+            {
+                hpx::execution::experimental::set_done(std::move(r));
+            }
+
+        private:
+            template <typename... Ts>
+            void set_value_helper(std::true_type, Ts&&... ts) noexcept
+            {
+                hpx::detail::try_catch_exception_ptr(
+                    [&]() {
+                        for (auto const& s : shape)
+                        {
+                            HPX_INVOKE(f, s, ts...);
+                        }
+                        hpx::execution::experimental::set_value(std::move(r));
+                    },
+                    [&](std::exception_ptr ep) {
+                        hpx::execution::experimental::set_error(
+                            std::move(r), std::move(ep));
+                    });
+            }
+
+            template <typename... Ts>
+            void set_value_helper(std::false_type, Ts&&... ts) noexcept
+            {
+                hpx::detail::try_catch_exception_ptr(
+                    [&]() {
+                        using result_type =
+                            hpx::util::invoke_result_t<std::decay_t<F>,
+                                shape_element, std::decay_t<Ts>...>;
+
+                        std::vector<result_type> results;
+                        results.reserve(util::size(shape));
+                        for (auto const& s : shape)
+                        {
+                            results.emplace_back(HPX_INVOKE(f, s, ts...));
+                        }
+
+                        hpx::execution::experimental::set_value(
+                            std::move(r), std::move(results));
+                    },
+                    [&](std::exception_ptr ep) {
+                        hpx::execution::experimental::set_error(
+                            std::move(r), std::move(ep));
+                    });
+            }
+
+        public:
+            template <typename... Ts,
+                typename = std::enable_if_t<
+                    hpx::is_invocable_v<F, shape_element, Ts...>>>
+            void set_value(Ts&&... ts) && noexcept
+            {
+                using is_void_result = std::is_void<
+                    hpx::util::invoke_result_t<F, shape_element, Ts...>>;
+                set_value_helper(is_void_result{}, std::forward<Ts>(ts)...);
```

Comment:
Ok, you're right, I misread this. I would like to have another facility in this case (`bulk_transform`?) that actually sends the result of the bulk execution.

Reply:
FWIW, I don't understand the rationale of why `bulk` is specified this way. This is rather ... surprising. Do you have any insights?

Reply:
I can't read the author's minds, but my guess is performance, plus you can build all kinds of primitives on top a `bulk` that's defined like this.

```
mycontainer in, out;
just(in, out) | bulk(n, [](int i, mycontainer& in, mycontainer& out) { out[i] = f(in[i]); }) | ...;
```

Additionally, defining `bulk` as it is right now hardcodes the use of `vector` for the output.

Reply:
> Additionally, defining `bulk` as it is right now hardcodes the use of `vector` for the output.

Yes, I was planning to change it to sending the produced values to the successor one by one.

Reply:
That's not allowed, at least the way things are specified right now. If I remember correctly the libunifex people had something in that direction at some point in some proposal, but not in P2300 and it requires the receiver being able to receive multiple values, and knowing when it has received the final value.

Reply:
Ok, my understanding was that one could call `set_value` more than once if it is followed by a `set_done`. Otherwise, why would we need `set_done` in the first place?

Reply:
`set_done` is meant for cancellation (the name is perhaps worth some bikeshedding on the mailing lists given this misunderstanding...). None of the current sender implementations (ours, not in general) support cancellation except by just forwarding the `set_done` signal from predecessors.

When any of the `set_*` functions is called, the receiver contract is fulfilled.

Reply:
Thanks for the explanations. In any case, I have addressed your comments, please re-review.